src/signature/spki.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright 2015 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR
+// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+use der;
+use signature;
+use untrusted;
+
+/// An error that occurs during certificate validation or name validation.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum VerifyWithSPKIError {
+    /// The encoding of some ASN.1 DER-encoded item is invalid.
+    BadDER,
+
+    /// The signature is invalid for the given public key.
+    InvalidSignatureForPublicKey,
+
+    /// The SignatureAlgorithm does not match the algorithm of the SPKI.
+    /// A mismatch could be because of the algorithm (RSA vs DSA, etc) or the
+    /// parameters (ECDSA_p256 vs ECDSA_384, etc).
+    UnsupportedSignatureAlgorithmForPublicKey,
+}
+
+/// Verify the signature `signature` of message `msg` with the public key
+/// `spki_public_key` using the algorithm `alg`.
+/// `signature::verify` is suitable when the public key bytes you have are the
+/// DER ASN.1 of the key. This method is suitable when the public key is in
+/// SubjectPublicKeyInfo form (https://tools.ietf.org/html/rfc5280#section-4.1),
+/// which is a wrapper containing both the key and an identifier.
+///
+/// A common situation where this encoding is encountered is when using public keys
+/// exported by OpenSSL. If you export an RSA or ECDSA public key from a keypair
+/// with `-pubout` and friends, you will get DER of an SPKI wrapper containing
+/// a public key. You could extract the bitstring of the key itself and use
+/// `signature::verify`, but that is often inconvient.
+pub fn verify(signature_alg: &Algorithm,
```

Comment:
Is it common that users would have some public key ASN.1 but not know what kind of key it was? That's not my use case but I don't claim to be representative.

It also complicates error handling somewhat. We would need an implicit priority of errors: presumably a `UnsupportedSignatureAlgorithmForPublicKey` would be suppressed, *unless* the last supplied algorithm produced that error, *or* an algorithm produced `InvalidSignatureForPublicKey`.

Anyway, for users that do have a bag o' algorithms that they wish to support, suppose we have this prelude:

```
let algorithms = &[ECDSA_P384_SHA384, ...];
let pub_key = ...;
let msg = ...;
let signature = ...;
```

Is the API change providing anything more than avoiding the `iter()` and `any()` here?

```
let got_a_live_one = 
    algorithms.iter().any(|a| spki::verify(a, pub_key, msg, signature).is_ok());
```

Or, if you wanted the algorithm, `filter()` and `nth(0)` instead of `any()`. To me, this seems like not an onerous burden for the consumer, and the intent is pretty evident I think. While it would be a little bit less code with the proposed changes, it would also require the user to look up API docs to determine how the `supported_algorithms` parameter is used (first match wins?) and how error handling works, whereas when the iteration is visible the semantics are pretty clear.

Anyway, I'll do it if you want, but IMO the syntactic win is not worth the semantic complexity.